src/java/org/apache/cassandra/service/reads/ReadCallback.java

```diff
@@ -160,24 +202,16 @@ public void awaitResults() throws ReadFailureException, ReadTimeoutException
          */
         int received = resolver.responses.size();
         boolean failed = failures > 0 && (blockFor > received || !resolver.isDataPresent());
-        WarningCounter warnings = warningCounter;
+        WarningContext warnings = warningContext;
+        // save the snapshot so abort state is not changed between now and when mayAbort gets called
+        TrackWarningsSnapshot snapshot = null;
         if (warnings != null)
         {
-            if (warnings.tombstoneAborts.get() > 0)
-            {
-                String msg = tombstoneAbortMessage(warnings.tombstoneAborts.get(), warnings.maxTombstoneAbortsCount.get(), command.toCQLString());
-                ClientWarn.instance.warn(msg + " with " + command.loggableTokens());
-                logger.warn(msg);
-                cfs().metric.clientTombstoneAborts.mark();
-            }
-
-            if (warnings.tombstoneWarnings.get() > 0)
-            {
-                String msg = tombstoneWarnMessage(warnings.tombstoneWarnings.get(), warnings.maxTombstoneWarningCount.get(), command.toCQLString());
-                ClientWarn.instance.warn(msg + " with " + command.loggableTokens());
-                logger.warn(msg);
-                cfs().metric.clientTombstoneWarnings.mark();
-            }
+            snapshot = warnings.snapshot();
+            // this is possible due to race condition
+            // network thread creates the WarningContext to update metrics, but we are activlly reading and see it is empty
```

Comment:
```suggestion
            // network thread creates the WarningContext to update metrics, but we are actively reading and see it is empty
```

Reply:
Would this only happen if the `WarningContext` is created by a responding thread (which contains warnings) after enough other threads respond (without warnings) to signal? In that case we would basically just ignore the warnings? (I think this could happen w/ speculative reads?)

Reply:
updated documents.  Speculative reads could cause this, but so could timeouts (signal=false)